aura.grammar

```diff
@@ -0,0 +1,95 @@
+# Note that DEDENT and INDENT are tokens which stand for spaces or tabs of a
+# consistent length. This makes that grammar context sensitive, so it cannot
+# actually be described in BNF.
+
+NAME: [A-Za-z][A-Xa-z0-9]*
+NEWLINE: '\n'
+
+use_statement: 'use' identifier 'as' identifier
+             | 'use' identifier
+
+function_def: 'fn' NAME '(' decl_list ')' type NEWLINE function_body
+
+return_statement: 'ret' expression
+                | 'ret'
+
+decl_list: decl | decl ',' decl_list | epsilon
+
+value: identifier | literal
+
+array: '[' value_list ']'
+
+tuple: '(' value_list ')'
+
+assignment_statement: identifier type assignment_op expression
+                    | '!' assignment_statement
+
+if_statement: 'if' expression NEWLINE INDENT body DEDENT
+
+switch_statement: 'switch' identifier cases
+
+case_body: statement
+         | 'next'
+		 | NEWLINE INDENT statement_or_next+ DEDENT
+
+statement_or_next: statement | next
+
+case: 'case' expression case_body
+    | 'case' expression
+    | 'default' expression case_body
+    | 'default' expression
+
+cases: NEWLINE INDENT case+ DEDENT
+
+identifier: NAME
+          | identifier '.' identifier
+
+value_list: epsilon
+          | value
+          | value ',' value
+
+
+behavior: 'on' NAME 'be' NAME functions
+        | 'on' NAME functions
+
+trait: 'trait' NAME function_decls
+
+struct: 'struct' NAME decls
+
+function_call: identifier '(' identifier_list ')'
+
+body: statement | NEWLINE INDENT statement+ DEDENT
+
+functions: NEWLINE INDENT function_def+ DEDENT
+
+decls: NEWLINE INDENT decl+ DEDENT
+
+decl: NAME type
+
+statement: if_statement | return_statement | function_def | behavior | trait
+         | struct | switch_statement | expression | assignment_statement
+         | use_statement
```

Comment:
A few of these aren't statements per se; since `body` has `statements+` then these don't belong here:

- `function_def` (though function *expressions* are valid, but those would be in an expression rule)
- `behavior`
- `trait`
- `struct`
- `use_statement` (questionable - let's leave it out for now)